### PR TITLE
Added graphviz visualization for obligation forests.

### PR DESCRIPTION
**Pull request by orium**
This can be a big help when debugging the trait resolver.

### CODE REVIEW

File: src/librustc_data_structures/obligation_forest/graphviz.rs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    ///
+    /// Calling this will do nothing unless the environment variable
+    /// `DUMP_OBLIGATION_FOREST_GRAPHVIZ` is defined.
+    ///
+    /// A few post-processing that you might want to do make the forest easier to visualize:
+    ///
+    ///  * `sed 's,std::[a-z]*::,,g'` — Deletes the `std::<package>::` prefix of paths.
+    ///  * `sed 's,"Binder(TraitPredicate(<\(.*\)>)) (\([^)]*\))","\1 (\2)",'` — Transforms
+    ///    `Binder(TraitPredicate(<predicate>))` into just `<predicate>`.
+    #[allow(dead_code)]
```
Comment (varkor) [inline src/librustc_data_structures/obligation_forest/graphviz.rs]: Why is this necessary?

Comment (orium) [reply src/librustc_data_structures/obligation_forest/graphviz.rs]: This function is not used.  The idea is that people can temporarily call it while debugging stuff.

**Comment by varkor** [comment]
I've not worked with `obligation_forest` before, so I think it'd be more helpful for someone who has to judge whether there are any ways this could be more useful. r? @arielb1

**Comment by arielb1** [comment]
Hi @orium - are you trying to debug something in particular?

**Comment by memoryruins** [comment]
@arielb1 

> (I'm glad I added graphviz to the obligations forest. It make things much easier to understand. I will open a PR soon with that.)

https://github.com/rust-lang/rust/pull/53255#issuecomment-421184780

**Comment by arielb1** [comment]
This looks like an OK idea, but I don't know the graphviz API that well - or what's the best way to integrate tooling like that into the compiler - r? @nikomatsakis for that.

**Comment by nikomatsakis** [comment]
I have no objection to landing this code, though I don't expect the obligation forest to live much longer.

**Comment by nikomatsakis** [comment]
The code itself seems fine :)

**Comment by orium** [comment]
@nikomatsakis rebased :)

**Comment by orium** [comment]
@nikomatsakis Should be mergeable now.  Btw, is there any way to avoid or reduce merge conflicts because of `Cargo.lock`?

**Comment by nikomatsakis** [comment]
@bors r+

@orium not that I know of
